### PR TITLE
feat(train): add optional `accumulate_grad_batches` config param

### PR DESCRIPTION
**Pull request by guranon**
Add an optional `accumulate_grad_batches` param to the `train` part of the config to allow for gradient accumulation. ([Lightning docs](https://lightning.ai/docs/pytorch/stable/common/optimization.html#gradient-accumulation))
This updates the gradients once every `accumulate_grad_batches` batches, with a default value of 1 to not break any existing configs.

I went with the same name (`accumulate_grad_batches`) as the [argument to `Trainer`](https://lightning.ai/docs/pytorch/stable/common/trainer.html#accumulate-grad-batches), even though we're not actually using that argument due to doing manual optimization. I realize that might be confusing, so I'm open to suggestions.

I'm relatively new to ML/AI work, so just to double-check my understanding:
- `manual_backward` backpropagates and accumulates gradients
- `step` updates model parameters
- `zero_grad` clears the gradients

So calling `manual_backward`, `step`, `zero_grad` should be equivalent to calling `zero_grad`, `manual_backward`, `step`.

Other considerations:
- Normalize/scale loss by 1 / accumulation steps
  - I looked at [NVIDIA's HiFi-GAN for PyTorch example](https://github.com/NVIDIA/DeepLearningExamples/blob/a6c678ef0323cdaae053e9eed34dce94488ad313/PyTorch/SpeechSynthesis/HiFiGAN/train.py#L389-L390) and they normalize the loss, so I'm pretty sure this is correct
- [We don't need to set `retain_graph=True` because we backpropagate once for each forward pass](https://stackoverflow.com/a/70119359)

### CODE REVIEW

**Comment by BlueAmulet** [comment]
One thing to consider is where gradient clipping should occur, my understanding is with gradient accumulation, clip_grad_value_ should also only be called once every N batches. Which on a slight note is probably broken since the training refactor, clip_grad_value_ is no longer after backwards and before optimizer step.

**Comment by 34j** [comment]
@allcontributors add BlueAmulet maintenance

**Comment by 34j** [comment]
@allcontributors add GarrettConway review
